File: Behavioral/TemplateMethod/CSharp/DataProcessor.cs
// DataProcessor.cs

using System;

public abstract class DataProcessor {
  // Template method defining the structure of the algorithm
  public void Process() {
    ReadData();
    ProcessData();
    SaveData();
  }

  // Abstract method to be implemented by subclasses
  protected abstract void ReadData();

  // Concrete method with common behavior
  protected void ProcessData() {
    Console.WriteLine("Processing data...");
  }

  // Concrete method with common behavior
  protected void SaveData() {
    Console.WriteLine("Data saved successfully.");
  }
}

File: Behavioral/TemplateMethod/Java/DataProcessor.java
// DataProcessor.java

public abstract class DataProcessor {
  
  // Template method defining the structure of the algorithm
  public final void process() {
    readData();
    processData();
    saveData();
  }

  // Abstract method to be implemented by subclasses
  protected abstract void readData();

  // Concrete method with common behavior
  protected void processData() {
    System.out.println("Processing data...");
  }

  // Concrete method with common behavior
  protected void saveData() {
    System.out.println("Data saved successfully.");
  }
}
